### PR TITLE
[WIP] fix: Using search backend for OR queries

### LINKED ISSUE #928: Search pagination is not returning all results
Sometimes there is an issue with the documents returned at the last page. We need to debug if it is server issue or client iteration issue.

### LINKED ISSUE #979: Paginated search retrieves incomplete documents upon multiple filters
The following code runs on each request for a specifc `page` number:
```ts
  const searchPaginatedData = await invoicesCollection.search(
    {
      q: '',
      searchFields: [
        "invoiceNumber",
        "customer.customerName",
        "customer.customerPhone",
      ],
      filter: {
        op: LogicalOperator.OR,
        selectorFilters: [
          { status: "Fully Paid" },
          { status: "Partial Paid" },
          { status: "Unpaid" },
        ],
      },
      hitsPerPage: 5,
      sort: { field: "createdAt", order: Order.DESC },
    },
    page
  );
  console.log("Hits count", searchPaginatedData.hits.length);
  console.log(searchPaginatedData.meta);
```
Following are the console outputs for 3 requests, one for each page. The first request shows correct document/hits count, but the subsequent pages are missing data:
1. `page = 1` All good as 5 documents retrieved for a page size of 5
```
Hits count 5
SearchMeta {
  _found: 11,
  _totalPages: 3,
  _page: Page { _current: 1, _size: 5 },
  _matchedFields: []
}
```
2. `page = 2` Only 1 document retrieved when there should have been 5
```
Hits count 1
SearchMeta {
  _found: 11,
  _totalPages: 3,
  _page: Page { _current: 2, _size: 5 },
  _matchedFields: []
}
```
3. `page = 3` Zero documents retrieved when there should have been 1
```
Hits count 0
SearchMeta {
  _found: 11,
  _totalPages: 3,
  _page: Page { _current: 3, _size: 5 },
  _matchedFields: []
}
```
*Note*: The issue only arises if I have more than one filters combined via logical OR. Seems to be working fine with a single filter or no filters

### PR DESCRIPTION
**Pull request by adilansari**
## Describe your changes
- Tigris was making multiple requests to search backend when `OR` filters were used and results were being merged from those requests in a sorted manner.
- This change utilizes `OR` filters in the search backend itself. Tigris now only sends a single search request, no more merging and sorting of results

## How best to test these changes
- Unit tests
- curl queries on a collection

## Issue ticket number and link
closes #979

### CODE REVIEW

File: query/filter/filter.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ type Filter interface {
 	Matches(doc []byte) bool
 	// MatchesDoc similar to Matches but used when document is already parsed
 	MatchesDoc(doc map[string]interface{}) bool
-	ToSearchFilter() []string
```
Comment (adilansari) [inline query/filter/filter.go:54]: We are not sending multiple filters to search backend anymore. A single string filter for all and/or queries.

File: query/filter/logical.go [hunk 2]
```diff
@@ -101,55 +103,8 @@ func (a *AndFilter) String() string {
 	return str + "}"
 }
 
-func (a *AndFilter) ToSearchFilter() []string {
```
Comment (adilansari) [inline query/filter/logical.go:104]: this is refactored into `searchSerializer` struct

File: query/filter/logical_test.go [hunk 3]
```diff
@@ -40,43 +40,38 @@ func TestLogicalToSearch(t *testing.T) {
 	}
 
 	js := []byte(`{"f1": 10}`)
-	testLogicalSearch(t, js, factory, []string{"f1:=10"})
+	testLogicalSearch(t, js, factory, "f1:=10")
```
Comment (adilansari) [inline query/filter/logical_test.go:43]: This test demonstrates changes to search filter serialization for the search backend

File: server/search/facets.go [hunk 4]
```diff
@@ -15,186 +15,84 @@
...
 	api "github.com/tigrisdata/tigris/api/server/v1"
-	"github.com/tigrisdata/tigris/lib/container"
+	"github.com/tigrisdata/tigris/query/search"
 	tsApi "github.com/typesense/typesense-go/typesense/api"
 )
 
-// SortedFacets is a Temporary workaround to merge facet values when aggregating results from
-// multi-search queries resulting from OR filters
-// this is not very efficient as of now.
-type SortedFacets struct {
```
Comment (adilansari) [inline server/search/facets.go:28]: No more sorting and merging of facets

File: server/search/factory.go [hunk 5]
```diff
@@ -61,12 +61,7 @@ func (r *ResponseFactory) GetGroupedHitsIterator(response []tsApi.SearchResult)
 
 // GetHitsIterator returns an IHits interface which contains hits results in an order that we need to stream out to the user.
 func (r *ResponseFactory) GetHitsIterator(response []tsApi.SearchResult) IHits {
-	var hits IHitsMutable
-	if r.inputQuery.SortOrder != nil && len(r.inputQuery.WrappedF.SearchFilter()) > 1 {
```
Comment (adilansari) [inline server/search/factory.go:65]: No need to sort hits, the search backend will return them sorted

**Review by himank — APPROVED** [review]

**Pull request merged**